### PR TITLE
Downgraded Python version to 3.12.3

### LINKED ISSUE #2330: [Bug]: "Publish Docker Image" CI workflow is failing
### Is there an existing issue for the same bug?

- [X] I have checked the troubleshooting document at https://opendevin.github.io/OpenDevin/modules/usage/troubleshooting
- [X] I have checked the existing issues.

### Describe the bug

Currently docker build seems to be consistently failing, one example here: https://github.com/OpenDevin/OpenDevin/actions/runs/9394171049/job/25973749341?pr=2081

### Current OpenDevin version

```bash
main
```

### Installation and Configuration

```bash
n/a
```

### Model and Agent

_No response_

### Operating System

_No response_

### Reproduction Steps

_No response_

### Logs, Errors, Screenshots, and Additional Context

_No response_

### PR DESCRIPTION
**Pull request by SmartManoj**
Temporarily fixes #2330

Consistent with #885

### CODE REVIEW

**Comment by tobitege** [comment]
3.12.3 might be better https://github.com/pydantic/pydantic/issues/9609#issuecomment-2155832461

**Comment by SmartManoj** [comment]
> Consistent with #885

For non-docker method, we are currently using 3.11 only.

**Comment by enyst** [comment]
In my recollection, we were using 3.12, when eventually we found out too many users were running into an issue with a Chroma dependency on 3.12, and we have downgraded for that reason, temporarily.

I would personally favor keeping 3.12 as much as we can, and go the other way around: hope to fix issues that prevent the upgrade, rather than continue downgrading.

That said ... it is a bit weird at this time. 😅

**Comment by neubig** [comment]
If 3.12.3 solves the problem I generally prefer going with newer versions of things myself.

File: containers/app/Dockerfile [hunk 1]
```diff
@@ -10,7 +10,7 @@ RUN npm ci
 COPY ./frontend ./
 RUN npm run make-i18n && npm run build
 
-FROM python:3.12-slim as backend-builder
+FROM python:3.11-slim as backend-builder
```
Comment (enyst) [inline containers/app/Dockerfile:13]:
```suggestion
FROM python:3.12.3-slim as backend-builder
```

File: containers/app/Dockerfile [hunk 2]
```diff
@@ -28,7 +28,7 @@ COPY ./pyproject.toml ./poetry.lock ./
 RUN touch README.md
 RUN poetry install --without evaluation --no-root && rm -rf $POETRY_CACHE_DIR
 
-FROM python:3.12-slim as runtime
+FROM python:3.11-slim as runtime
```
Comment (enyst) [inline containers/app/Dockerfile:31]:
```suggestion
FROM python:3.12.3-slim as runtime
```

**Comment by enyst** [comment]
My fault for the confusion. I did think our mix of python versions was a bit too funny. But this is the full `app` image that people run, and it works as it is.

This is only an issue with building it, and there's a PR on pydantic so it will be fixed upstream.

If we're okay with trying 3.12.3, this is a successful test for the edits suggested:
https://github.com/enyst/OpenDevin/actions/runs/9431038477/job/25979245963?pr=6

**Comment by SmartManoj** [comment]
https://github.com/OpenDevin/OpenDevin/pull/1664#discussion_r1632159141

**Review by xingyaoww — APPROVED** [review]
LGTM

**Pull request merged**